llvm/lib/CAS/ObjectStore.cpp

```diff
@@ -98,6 +99,48 @@ Expected<ObjectProxy> ObjectStore::createProxy(ArrayRef<ObjectRef> Refs,
   return getProxy(*Ref);
 }
 
+Expected<ObjectRef>
+ObjectStore::storeFromOpenFileImpl(sys::fs::file_t FD,
+                                   Optional<sys::fs::file_status> Status) {
+  // \p Copy the file into an immutable memory buffer and call \p store on that.
+  // Using \p mmap would be unsafe because there's a race window between when we
+  // get the digest hash for the \p mmap contents and when we store the data; if
+  // the file changes in-between we will create an invalid object.
+
+  // FIXME: For the on-disk CAS implementation use cloning to store it as a
+  // standalone file if the file-system supports it and the file is large.
+
+  constexpr size_t ChunkSize = 4 * 4096;
+  // FIXME: MSVC: SmallString<ChunkSize * 2>
+  SmallString<4 * 4096 * 2> Data;
```

Comment:
This is a massive stack buffer; is there a strong reason to not just do `SmallString<0>` and `reserve(ChunkSize * 2)` on the heap?  I realize this is code you moved from elsewhere, but I hadn't noticed it before.

Reply:
Good idea, done.